Revise and expand on 'Component Model Concepts' section and its subsections (#286)
bytecodealliance/component-docs: component-model/src/design/component-model-concepts.md

## Component Model Concepts

The WebAssembly Component Model extends core WebAssembly
by introducing higher-level types
and interface-driven development.
The Component Model makes core WebAssembly composable:
components that provide functionality and those that use them

> can be composed together into *one* resulting component.

Author: How does [97519f5](https://github.com/bytecodealliance/component-docs/pull/286/commits/97519f5329eaafba71d2f1403f9d1d8a36c1eb2f) look? Is there anything else you would want to add to the list?

This section introduces the core concepts behind the component model.
For the rationale behind the component model, see [the previous section](./why-component-model.md).

### Components

> A [WebAssembly Component](./components.md) is a core module extended with higher-level types and interfaces.

Reviewer: OK looks great! Spacing is a bit weird IMO *but* we can cover that some other time I think. I've filed an issue so we can get our ducks in a row on the format we want/expect for markdown.

WebAssembly components are *nestable*:
they may contain one or more core modules and/or sub-components composed together.
For example, a component implementing a simple calculator might be written
by composing together a component that parses strings to floating-point numbers
with a component that does the main arithmetic.

### WebAssembly Interface Types (WIT)

[WebAssembly Interface Types (WIT)][wit] is the [IDL (Interface Definition Language)][wiki-idl]
used to formally define functionality for WebAssembly modules.
WIT gives WebAssembly components the ability to express type signatures
in a language-agnostic way,
so any component binary can be both checked and executed.

### Interfaces

An [_interface_](./interfaces.md) is a collection of type definitions
and function declarations (function names accompanied by type signatures).
Typically, a single interface describes a specific, focused bit
of functionality.
For example, in [wasi-cli][wasi-cli-stdio],
three separate interfaces are used to implement `stdin`, `stdout`, and `stderr`
(the three input and output streams typically available
in a command-line environment.)

### Worlds

> A [_world_](./worlds.md) is a collection of interfaces

Reviewer: ```suggestion
A [_world_](./worlds.md) is a collection of interfaces and types
```

including types in worlds is bad practice (interfaces are the ideal), but probably worth mentioning here since we mention it in other sections.

that expresses what features a component offers
and what features it depends on.

> For example, wasi-cli includes the [`stdio` world][wasi-cli-stdio],

Reviewer: Ah so the `stdio` isn't  a world but [`command` is](https://github.com/WebAssembly/wasi-cli/blob/1412e7b3f211f50a9f642c9b765b0a53d9b16e74/wit/command.wit#L4)! This section could be reworked to use that as an example instead. (and it's probably a better one), building on the previous explanations and how close it is to something practical

which collects together three separate interfaces
that represent the `stdin`, `stdout`, and `stderr` streams.
Any component implementing the `stdio` world
must implement those three interfaces.

### Packages

 A [_package_](./packages.md) is a set of WIT files
 containing a related set of interfaces and worlds.
 For example, the wasi-cli package includes
 the `stdio` world and the `environment` world, among others,

>  with each defined in its own WIT file.

Reviewer: It might be worthwhile to show how packages actually show up in WIT files here? Or link out to wasi-cli/wit? Not sure if anything can be done here since it feels like we want to keep this section abstract/not get into the literal syntax too much. 

Oh also `stdio` not being a valid world (it ends up just being a file in the folder that gets picked up because of resolution rules), and neither is `environment` (it's an interface)

Author: Fixed in [0c1d9a7](https://github.com/bytecodealliance/component-docs/pull/286/commits/0c1d9a792a41e013654d4eb51796e820a65b9be4)

### Platforms

In the context of WebAssembly, the _host_ is the browser or stand-alone runtime
that runs WebAssembly modules.

> The _guest_ is the WebAssembly module that is executed by the host.

Reviewer: ```suggestion
A _guest_ refers to the WebAssembly binary that is executed by the host.
```

> (These terms come from virtualization, where a guest is

Reviewer: ```suggestion
(These terms borrow from their analogs in [virtualization](https://en.wikipedia.org/wiki/Virtualization), where a guest is
```
(If you decide to take this, feel free to move that link down!)

a software-based virtual machine that runs on physical hardware,
which is the "host")

The Component Model introduces the idea of a _platform_
to core WebAssembly—enabling the structured, standardized use

> of host functionality for WebAssembly guests.

Reviewer: We might benefit from maybe *one* more sentence that describes how this works -- i.e. components may `import` functionality that is fulfilled by the platform underneath them.

## WASI

The WebAssembly System Interface ([WASI][wasi]) defines in WIT
a family of interfaces for common system-level functions.
WASI standardizes the functionality provided by platforms,
so that component writers can rely on functionality
that is guaranteed to be available on any conformant platform.

WASI defines common collections of functionality
such as the command line ([`wasi:cli`][wasi-cli])

> or an HTTP server ([`wasi:http`][wasi-http]).

Reviewer: ```suggestion
or an HTTP servers and clients ([`wasi:http`][wasi-http]).
```

Assuming you don't take suggestion above this one and combine it all together.

Author: Deleted this sentence altogether since it's now covered in the previous para.

## Component Model Concepts

The WebAssembly Component Model extends core WebAssembly in several ways.
The Component Model:
* Adds consistent representation of higher-level types
* Enables interface-driven development
* Makes core WebAssembly composable:
components that provide functionality and those that use them
can be composed together into *one* resulting component.

This section introduces the core concepts behind the component model.
For the rationale behind the component model, see [the previous section](./why-component-model.md).

### Components

A [WebAssembly Component](./components.md) is a binary that
conforms to the [Canonical ABI](../advanced/canonical-abi.md);
often a WebAssembly core module extended with the features
of the Component Model
(higher-level types, interfaces).
WebAssembly components are *nestable*:
they may contain zero or more core modules and/or sub-components composed together.
For example, a component implementing a simple calculator might be written
by composing together a component that parses strings to floating-point numbers
with a component that does the main arithmetic.

### WebAssembly Interface Types (WIT)

> [WebAssembly Interface Types (WIT)][wit] is the [IDL (Interface Definition Language)][wiki-idl]

Reviewer: ```suggestion
[WebAssembly Interface Types (WIT)][wit] is the [Interface Definition Language (IDL)][wiki-idl]
```

used to formally define functionality for WebAssembly components.
WIT gives WebAssembly components the ability to express type signatures
in a language-agnostic way,
so any component binary can be checked, composed and executed.

#### Interfaces

An [_interface_](./interfaces.md) is a collection of type definitions
and function declarations (function names accompanied by type signatures).
Typically, a single interface describes a specific, focused bit

> of functionality.
> For example, in [wasi-cli][wasi-cli-stdio],

Reviewer: ```suggestion
of functionality.

For example, in [wasi-cli][wasi-cli-stdio],
```

three separate interfaces are used to implement `stdin`, `stdout`, and `stderr`
(streams typically available in command-line-like environments)

### Worlds

A [_world_](./worlds.md) is a collection of interfaces and types
that expresses what features a component offers
and what features it depends on.
For example, wasi-cli includes the [`command` world][wasi-cli-command],
which depends on interfaces
that represent the `stdin`, `stdout`, and `stderr` streams,
among other things.
A component implementing the `command` world
must be invoked in an environment that implements those interfaces.

### Packages

 A [_package_](./packages.md) is a set of WIT files

>  containing a related set of interfaces and worlds.
>  For example, the [wasi-http](https://github.com/WebAssembly/wasi-http/blob/main/wit/proxy.wit) package includes

Reviewer: ```suggestion
containing a related set of interfaces and worlds.
 
For example, the [wasi-http](https://github.com/WebAssembly/wasi-http/blob/main/wit/proxy.wit) package includes
```
